File: src/minigames/paintball/java/work/lclpnet/ap2/game/paintball/item/TripWireItem.java
package work.lclpnet.ap2.game.paintball.item;

import net.minecraft.ChatFormatting;
import net.minecraft.core.RegistryAccess;
import net.minecraft.core.particles.DustParticleOptions;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.level.ClipContext;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.EntityHitResult;
import net.minecraft.world.phys.HitResult;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.phys.shapes.CollisionContext;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.ap2.api.base.Participants;
import work.lclpnet.ap2.api.game.team.Team;
import work.lclpnet.ap2.game.paintball.util.PaintManager;
import work.lclpnet.ap2.game.paintball.util.PaintballTeam;
import work.lclpnet.ap2.game.paintball.util.PaintballTeams;
import work.lclpnet.ap2.impl.game.item.SpecialItem;
import work.lclpnet.ap2.impl.game.item.SpecialItemContext;
import work.lclpnet.kibu.scheduler.api.TaskScheduler;
import work.lclpnet.kibu.translate.Translations;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

import static work.lclpnet.ap2.impl.util.ParticleHelper.spawnParticleFor;
import static work.lclpnet.ap2.impl.util.RayCastUtil.*;
import static work.lclpnet.ap2.impl.util.SoundHelper.playSound;
import static work.lclpnet.ap2.impl.util.SoundHelper.playSoundFor;

public class TripWireItem implements SpecialItem {

    private static final double
            MAX_LENGTH = 7,
            TRIPWIRE_MARGIN = 0.01;

    private final Translations translations;
    private final Participants participants;
    private final ServerLevel world;
    private final PaintballTeams teams;
    private final PaintManager paintManager;
    private final Set<Tripwire> tripwires = new HashSet<>();

    public TripWireItem(Translations translations, Participants participants, ServerLevel world, PaintballTeams teams,
                        PaintManager paintManager) {
        this.translations = translations;
        this.participants = participants;
        this.world = world;
        this.teams = teams;
        this.paintManager = paintManager;
    }

    @Override
    public String id() {
        return "tripwire";
    }

    @Override
    public ItemStack createItemStack(RegistryAccess registryManager) {
        return new ItemStack(Items.TRIPWIRE_HOOK);
    }

    @Override
    public InteractionResult onUse(ServerPlayer player, ItemStack stack, @Nullable InteractionHand hand, SpecialItemContext ctx) {
        double range = player.getAttributeValue(Attributes.BLOCK_INTERACTION_RANGE);

        HitResult hit = raycast(player, range, ClipContext.Block.OUTLINE,
                ClipContext.Fluid.NONE, CollisionContext.empty(), entity -> !entity.isSpectator());

        if (hit.getType() != HitResult.Type.BLOCK || !(hit instanceof BlockHitResult blockHit)) {
            return InteractionResult.PASS;
        }

        Vec3 pos = blockHit.getLocation();
        Vec3 dir = blockHit.getDirection().getUnitVec3();

        BlockHitResult opposingHit = raycastBlocks(world, pos.add(dir.scale(TRIPWIRE_MARGIN)), dir, MAX_LENGTH,
                ClipContext.Block.OUTLINE, ClipContext.Fluid.NONE, CollisionContext.empty());

        if (opposingHit.getType() != HitResult.Type.BLOCK) {
            translations.translateText("game.ap2.paintball.item.tripwire.too_long")
                    .formatted(ChatFormatting.RED)
                    .sendTo(player);

            player.playNotifySound(SoundEvents.NOTE_BLOCK_BASS.value(), SoundSource.BLOCKS, 0.2f, 1f);

            return InteractionResult.FAIL;
        }

        stack.consume(1, player);

        double length = opposingHit.getLocation().subtract(pos).length();

        tripwires.add(new Tripwire(pos, dir, length, player.getUUID()));

        float activateVolume = 0.45f, activatePitch = 1.78f;
        float placeVolume = 0.5f, placePitch = 1.3f;

        teams.getTeamManager().getTeam(player).ifPresentOrElse(
                team -> {
                    playSoundFor(SoundEvents.BEACON_ACTIVATE, SoundSource.PLAYERS, pos, activateVolume, activatePitch, team.getPlayers());
                    playSoundFor(SoundEvents.IRON_PLACE, SoundSource.PLAYERS, pos, placeVolume, placePitch, team.getPlayers());
                },
                () -> {
                    playSound(player, SoundEvents.BEACON_ACTIVATE, SoundSource.PLAYERS, pos, activateVolume, activatePitch);
                    playSound(player, SoundEvents.BEACON_ACTIVATE, SoundSource.PLAYERS, pos, placeVolume, placePitch);
                }
        );

        return InteractionResult.SUCCESS_SERVER;
    }

    @Override
    public void scheduleTasks(TaskScheduler scheduler, SpecialItemContext ctx) {
        scheduler.interval(this::tick, 1);
    }

    private void tick() {
        tripwires.removeIf(Tripwire::tick);
    }

    private class Tripwire {

        private static final float
                DISPLAY_LASER_SPACING = 0.125f,
                DISPLAY_LASER_SIZE = 0.25f,
                DISPLAY_LASER_TICKS = 4,
                EXPLOSION_POWER = 3.5f;

        private final Vec3 pos;
        private final Vec3 dir;
        private final double length;
        private final UUID ownerUuid;
        private int timer = 0;

        private Tripwire(Vec3 pos, Vec3 dir, double length, UUID ownerUuid) {
            this.pos = pos;
            this.dir = dir;
            this.length = length;
            this.ownerUuid = ownerUuid;
        }

        public boolean tick() {
            ServerPlayer player = participants.getParticipant(ownerUuid).orElse(null);

            if (player == null) return true;

            Team team = teams.getTeamManager().getTeam(player).orElse(null);

            if (team == null) return true;

            PaintballTeam paintballTeam = teams.teamOf(player).orElse(null);

            if (paintballTeam == null) return true;

            if (timer++ % DISPLAY_LASER_TICKS == 0) {
                showTo(team);
            }

            return checkExplosion(player, paintballTeam);
        }

        private void showTo(Team team) {
            for (double d = 0; d <= length; d += DISPLAY_LASER_SPACING) {
                DustParticleOptions effect = new DustParticleOptions(team.key().color(), DISPLAY_LASER_SIZE);

                spawnParticleFor(effect, pos.x + dir.x * d, pos.y + dir.y * d, pos.z + dir.z * d,
                        1, 0, 0, 0, 0, team.getPlayers());
            }
        }

        private boolean checkExplosion(ServerPlayer owner, PaintballTeam ownerTeam) {
            HitResult hit = raycastEntities(world, pos.add(dir.scale(TRIPWIRE_MARGIN)), dir, length, entity
                    -> entity instanceof ServerPlayer player
                    && participants.isParticipating(player)
                    && teams.teamOf(player).map(pbt -> pbt.key() != ownerTeam.key()).orElse(false));

            if (hit.getType() != HitResult.Type.ENTITY || !(hit instanceof EntityHitResult entityHit)) return false;

            paintManager.createExplosion(owner, entityHit.getEntity().position(), ownerTeam, EXPLOSION_POWER);

            return true;
        }
    }
}

File: src/minigames/paintball/src/main/kotlin/work/lclpnet/ap2/game/paintball/item/TripWireItem.kt
package work.lclpnet.ap2.game.paintball.item

import net.minecraft.ChatFormatting
import net.minecraft.core.RegistryAccess
import net.minecraft.core.particles.DustParticleOptions
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.level.ClipContext
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.EntityHitResult
import net.minecraft.world.phys.HitResult
import net.minecraft.world.phys.Vec3
import net.minecraft.world.phys.shapes.CollisionContext
import work.lclpnet.ap2.ext.mc.playNotifySound
import work.lclpnet.ap2.game.paintball.util.PaintManager
import work.lclpnet.ap2.game.paintball.util.PaintballTeam
import work.lclpnet.ap2.game.paintball.util.PaintballTeams
import work.lclpnet.ap2.game.player.Participants
import work.lclpnet.ap2.game.team.Team
import work.lclpnet.ap2.impl.game.item.SpecialItem
import work.lclpnet.ap2.impl.game.item.SpecialItemContext
import work.lclpnet.ap2.impl.util.ParticleHelper.spawnParticleFor
import work.lclpnet.ap2.impl.util.RayCastUtil.*
import work.lclpnet.ap2.impl.util.SoundHelper.playSound
import work.lclpnet.ap2.impl.util.SoundHelper.playSoundFor
import work.lclpnet.kibu.scheduler.api.TaskScheduler
import work.lclpnet.kibu.translate.Translations
import java.util.*

private const val MAX_LENGTH = 7.0
private const val TRIPWIRE_MARGIN = 0.01

private const val DISPLAY_LASER_SPACING = 0.125f
private const val DISPLAY_LASER_SIZE = 0.25f
private const val DISPLAY_LASER_TICKS = 4
private const val TRIPWIRE_EXPLOSION_POWER = 3.5f

class TripWireItem(
    private val translations: Translations,
    private val participants: Participants,
    private val world: ServerLevel,
    private val teams: PaintballTeams,
    private val paintManager: PaintManager,
    private val onUsed: (ServerPlayer) -> Unit
) : SpecialItem {

    private val tripwires = HashSet<Tripwire>()

    override fun id() = "tripwire"

    override fun createItemStack(registryManager: RegistryAccess): ItemStack = ItemStack(Items.TRIPWIRE_HOOK)

    override fun onUse(player: ServerPlayer, stack: ItemStack, hand: InteractionHand?, ctx: SpecialItemContext): InteractionResult {
        val range = player.getAttributeValue(Attributes.BLOCK_INTERACTION_RANGE)

        val hit = raycast(player, range, ClipContext.Block.OUTLINE,
            ClipContext.Fluid.NONE, CollisionContext.empty()) { !it.isSpectator }

        if (hit.type != HitResult.Type.BLOCK || hit !is BlockHitResult) return InteractionResult.PASS

        val pos = hit.location
        val dir = hit.direction.unitVec3

        val opposingHit = raycastBlocks(
            world,
            pos.add(dir.scale(TRIPWIRE_MARGIN)),
            dir,
            MAX_LENGTH,
            ClipContext.Block.OUTLINE,
            ClipContext.Fluid.NONE,
            CollisionContext.empty()
        )

        if (opposingHit.type != HitResult.Type.BLOCK) {
            translations.translateText("item.tripwire.too_long")
                .withStyle(ChatFormatting.RED)
                .sendTo(player)

            player.playNotifySound(SoundEvents.NOTE_BLOCK_BASS.value(), SoundSource.BLOCKS, 0.2f, 1f)

            return InteractionResult.FAIL
        }

        stack.consume(1, player)

        onUsed(player)

        val length = opposingHit.location.subtract(pos).length()

        tripwires.add(Tripwire(pos, dir, length, player.uuid))

        val activateVolume = 0.45f
        val activatePitch = 1.78f
        val placeVolume = 0.5f
        val placePitch = 1.3f

        val team = teams.teamManager.getTeam(player)

        if (team != null) {
            playSoundFor(SoundEvents.BEACON_ACTIVATE, SoundSource.PLAYERS, pos, activateVolume, activatePitch, team.players)
            playSoundFor(SoundEvents.IRON_PLACE, SoundSource.PLAYERS, pos, placeVolume, placePitch, team.players)
        } else {
            playSound(player, SoundEvents.BEACON_ACTIVATE, SoundSource.PLAYERS, pos, activateVolume, activatePitch)
            playSound(player, SoundEvents.BEACON_ACTIVATE, SoundSource.PLAYERS, pos, placeVolume, placePitch)
        }

        return InteractionResult.SUCCESS_SERVER
    }

    override fun scheduleTasks(scheduler: TaskScheduler, ctx: SpecialItemContext) {
        scheduler.interval(1, ::tick)
    }

    private fun tick() {
        tripwires.removeIf(Tripwire::tick)
    }

    private inner class Tripwire(
        private val pos: Vec3,
        private val dir: Vec3,
        private val length: Double,
        private val ownerUuid: UUID
    ) {
        private var timer = 0

        fun tick(): Boolean {
            val player = participants.getParticipant(ownerUuid).orElse(null) ?: return true

            val team: Team = teams.teamManager.getTeam(player) ?: return true

            val paintballTeam: PaintballTeam = teams.teamOf(player) ?: return true

            if (timer++ % DISPLAY_LASER_TICKS == 0) {
                showTo(team)
            }

            return checkExplosion(player, paintballTeam)
        }

        private fun showTo(team: Team) {
            var d = 0.0

            while (d <= length) {
                val effect = DustParticleOptions(team.key.color, DISPLAY_LASER_SIZE)

                spawnParticleFor(effect, pos.x + dir.x * d, pos.y + dir.y * d, pos.z + dir.z * d,
                    1, 0.0, 0.0, 0.0, 0.0, team.players)

                d += DISPLAY_LASER_SPACING
            }
        }

        private fun checkExplosion(owner: ServerPlayer, ownerTeam: PaintballTeam): Boolean {
            val hit = raycastEntities(world, pos.add(dir.scale(TRIPWIRE_MARGIN)), dir, length) { entity ->
                entity is ServerPlayer
                    && participants.isParticipating(entity)
                    && teams.teamOf(entity)?.let { it.key != ownerTeam.key } ?: false
            }

            if (hit.type != HitResult.Type.ENTITY || hit !is EntityHitResult) return false

            paintManager.createExplosion(owner, hit.entity.position(), ownerTeam, TRIPWIRE_EXPLOSION_POWER)

            return true
        }
    }
}
